File: mage/src/main/java/mil/nga/giat/mage/newsfeed/ObservationListAdapter.java
package mil.nga.giat.mage.newsfeed;

import android.content.Context;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.text.format.DateUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.RecyclerView;

import com.j256.ormlite.android.AndroidDatabaseResults;
import com.j256.ormlite.stmt.PreparedQuery;

import java.lang.ref.WeakReference;
import java.sql.SQLException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Collection;
import java.util.Date;
import java.util.Locale;

import mil.nga.giat.mage.R;
import mil.nga.giat.mage.map.marker.ObservationBitmapFactory;
import mil.nga.giat.mage.observation.AttachmentGallery;
import mil.nga.giat.mage.sdk.datastore.observation.Observation;
import mil.nga.giat.mage.sdk.datastore.observation.ObservationError;
import mil.nga.giat.mage.sdk.datastore.observation.ObservationFavorite;
import mil.nga.giat.mage.sdk.datastore.observation.ObservationHelper;
import mil.nga.giat.mage.sdk.datastore.observation.ObservationImportant;
import mil.nga.giat.mage.sdk.datastore.observation.ObservationProperty;
import mil.nga.giat.mage.sdk.datastore.user.User;
import mil.nga.giat.mage.sdk.datastore.user.UserHelper;
import mil.nga.giat.mage.sdk.exceptions.ObservationException;
import mil.nga.giat.mage.sdk.exceptions.UserException;

/**
 * Created by wnewman on 10/30/17.
 */

public class ObservationListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public interface ObservationActionListener {
        void onObservationClick(Observation observation);
        void onObservationImportant(Observation observation);
        void onObservationDirections(Observation observation);
    }

    private static final String LOG_NAME = ObservationListAdapter.class.getName();

    private static final String SHORT_TIME_PATTERN = "h:mm a";
    private static final String SHORT_DATE_PATTERN = "MMM d";

    private static final int TYPE_OBSERVATION = 1;
    private static final int TYPE_FOOTER = 2;

    private Context context;
    private Cursor cursor;
    private PreparedQuery<Observation> query;
    private AttachmentGallery attachmentGallery;
    private User currentUser;
    private ObservationActionListener observationActionListener;
    private String footerText;

    private class ObservationViewHolder extends RecyclerView.ViewHolder {
        private ImageView markerView;
        private TextView primaryView;
        private TextView timeView;
        private TextView secondaryView;
        private TextView userView;
        private View importantView;
        private TextView importantOverline;
        private TextView importantDescription;
        private ImageView importantButton;
        private View syncBadge;
        private View errorBadge;
        private LinearLayout attachmentLayout;
        private ImageView favoriteButton;
        private TextView favoriteCount;
        private View directionsButton;
        private IconTask iconTask;
        private UserTask userTask;
        private PropertyTask primaryPropertyTask;
        private PropertyTask secondaryPropertyTask;

        ObservationViewHolder(View view) {
            super(view);

            markerView = view.findViewById(R.id.observation_marker);
            primaryView = view.findViewById(R.id.primary);
            timeView = view.findViewById(R.id.time);
            secondaryView = view.findViewById(R.id.secondary);
            userView = view.findViewById(R.id.user);
            importantView = view.findViewById(R.id.important);
            importantOverline = view.findViewById(R.id.important_overline);
            importantDescription = view.findViewById(R.id.important_description);
            importantButton = view.findViewById(R.id.important_button);
            syncBadge = view.findViewById(R.id.sync_status);
            errorBadge = view.findViewById(R.id.error_status);
            attachmentLayout = view.findViewById(R.id.image_gallery);
            favoriteButton = view.findViewById(R.id.favorite_button);
            favoriteCount = view.findViewById(R.id.favorite_count);
            directionsButton = view.findViewById(R.id.directions_button);
        }

        public void bind(final Observation observation, final ObservationActionListener listener) {
            itemView.setOnClickListener(v -> listener.onObservationClick(observation));
        }
    }

    private class FooterViewHolder extends RecyclerView.ViewHolder {
        private TextView footerText;

        FooterViewHolder(View view) {
            super(view);

            footerText = view.findViewById(R.id.footer_text);
        }
    }

    public ObservationListAdapter(Context context, AttachmentGallery attachmentGallery, ObservationActionListener observationActionListener) {
        this.context = context;
        this.attachmentGallery = attachmentGallery;
        this.observationActionListener = observationActionListener;
    }

    public void setCursor(Cursor cursor, PreparedQuery<Observation> query) {
        closeCursor();

        this.cursor = cursor;
        this.query = query;
        this.notifyDataSetChanged();
    }

    public void closeCursor() {
        if (cursor != null) {
            cursor.close();
            cursor = null;
        }
    }

    public void setFooterText(String footerText) {
        this.footerText = footerText;
    }

    @Override
    public int getItemViewType(int position) {
        if (cursor != null && position == cursor.getCount()) {
            return TYPE_FOOTER;
        } else {
            return TYPE_OBSERVATION;
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_OBSERVATION) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.observation_list_item, parent, false);
            return new ObservationViewHolder(itemView);
        } else if (viewType == TYPE_FOOTER) {
            View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.feed_footer, parent, false);
            return new FooterViewHolder(itemView);
        } else {
            return null;
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ObservationViewHolder) {
            bindObservationViewHolder(holder, position);
        } else if (holder instanceof FooterViewHolder) {
            bindFooterViewHolder(holder);
        }
    }

    @Override
    public int getItemCount() {
        if (cursor == null) return 0;

        return cursor.getCount() + 1;
    }

    @Override
    public void onViewRecycled(RecyclerView.ViewHolder holder) {
        if (holder instanceof ObservationViewHolder) {
            ObservationViewHolder vh = (ObservationViewHolder) holder;

            if (vh.iconTask != null) {
                vh.iconTask.cancel(false);
            }

            if (vh.userTask != null) {
                vh.userTask.cancel(false);
            }

            if (vh.primaryPropertyTask != null) {
                vh.primaryPropertyTask.cancel(false);
            }

            if (vh.secondaryPropertyTask != null) {
                vh.secondaryPropertyTask.cancel(false);
            }
        }
    }

    private void bindObservationViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (cursor == null) return;

        cursor.moveToPosition(position);
        ObservationViewHolder vh = (ObservationViewHolder) holder;

        try {
            final Observation observation = query.mapRow(new AndroidDatabaseResults(cursor, null, false));
            vh.bind(observation, observationActionListener);

            Drawable markerPlaceholder = DrawableCompat.wrap(ContextCompat.getDrawable(context, R.drawable.ic_place_white_48dp));
            DrawableCompat.setTint(markerPlaceholder, ContextCompat.getColor(context, R.color.icon));
            DrawableCompat.setTintMode(markerPlaceholder, PorterDuff.Mode.SRC_IN);

            vh.markerView.setImageDrawable(markerPlaceholder);
            vh.iconTask = new IconTask(vh.markerView);
            vh.iconTask.execute(observation);

            vh.primaryView.setText("");
            vh.primaryPropertyTask = new PropertyTask(PropertyTask.Type.PRIMARY, vh.primaryView);
            vh.primaryPropertyTask.execute(observation);

            vh.secondaryView.setText("");
            vh.secondaryPropertyTask = new PropertyTask(PropertyTask.Type.SECONDARY, vh.secondaryView);
            vh.secondaryPropertyTask.execute(observation);

            Date timestamp = observation.getTimestamp();
            String pattern = DateUtils.isToday(timestamp.getTime()) ? SHORT_TIME_PATTERN : SHORT_DATE_PATTERN;
            DateFormat dateFormat = new SimpleDateFormat(pattern, Locale.getDefault());
            vh.timeView.setText(dateFormat.format(timestamp));

            vh.userView.setText("");
            vh.userTask = new UserTask(vh.userView);
            vh.userTask.execute(observation);

            vh.importantButton.setOnClickListener(v -> observationActionListener.onObservationImportant(observation));
            setImportantView(observation.getImportant(), vh);

            ObservationError error = observation.getError();
            if (error != null) {
                boolean hasValidationError = error.getStatusCode() != null;
                vh.syncBadge.setVisibility(hasValidationError ? View.GONE : View.VISIBLE);
                vh.errorBadge.setVisibility(hasValidationError ? View.VISIBLE : View.GONE);
            } else {
                vh.syncBadge.setVisibility(View.GONE);
                vh.errorBadge.setVisibility(View.GONE);
            }

            vh.attachmentLayout.removeAllViews();
            if (observation.getAttachments().size() == 0) {
                vh.attachmentLayout.setVisibility(View.GONE);
            } else {
                vh.attachmentLayout.setVisibility(View.VISIBLE);
                attachmentGallery.addAttachments(vh.attachmentLayout, observation.getAttachments());
            }

            vh.favoriteButton.setOnClickListener(v -> toggleFavorite(observation, vh));
            setFavoriteImage(observation.getFavorites(), vh, isFavorite(observation));

            vh.directionsButton.setOnClickListener(v -> getDirections(observation));
        } catch (SQLException e) {
            e.printStackTrace();
        }
    }

    private void bindFooterViewHolder(RecyclerView.ViewHolder holder) {
        FooterViewHolder vh = (FooterViewHolder) holder;

        vh.footerText.setText(footerText);
    }

    private void setImportantView(ObservationImportant important, ObservationViewHolder vh) {
        boolean isImportant = important != null && important.isImportant();
        vh.importantView.setVisibility(isImportant ? View.VISIBLE : View.GONE);

        if (isImportant) {
            try {
                User user = UserHelper.getInstance(context).read(important.getUserId());
                vh.importantOverline.setText(String.format("FLAGGED BY %s", user.getDisplayName().toUpperCase()));
            } catch (UserException e) {
                e.printStackTrace();
            }

            vh.importantDescription.setText(important.getDescription());
        }

        if (isImportant) {
            vh.importantButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_flag_white_24dp));
            vh.importantButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_flag_active));
        } else {
            vh.importantButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_flag_outlined_white_24dp));
            vh.importantButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_flag_inactive));
        }
    }

    private void setFavoriteImage(Collection<ObservationFavorite> favorites, ObservationViewHolder vh, boolean isFavorite) {
        if (isFavorite) {
            vh.favoriteButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_favorite_white_24dp));
            vh.favoriteButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_favorite_active));
        } else {
            vh.favoriteButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_favorite_border_white_24dp));
            vh.favoriteButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_favorite_inactive));
        }

        vh.favoriteCount.setVisibility(favorites.size() > 0 ? View.VISIBLE : View.GONE);
        vh.favoriteCount.setText(String.format(Locale.getDefault(), "%d", favorites.size()));
    }

    private void toggleFavorite(Observation observation, ObservationViewHolder vh) {
        ObservationHelper observationHelper = ObservationHelper.getInstance(context);
        boolean isFavorite = isFavorite(observation);
        try {
            if (isFavorite) {
                observationHelper.unfavoriteObservation(observation, currentUser);
            } else {
                observationHelper.favoriteObservation(observation, currentUser);
            }

            setFavoriteImage(observation.getFavorites(), vh, isFavorite);
        } catch (ObservationException e) {
            Log.e(LOG_NAME, "Could not unfavorite observation", e);
        }
    }

    private boolean isFavorite(Observation observation) {
        boolean isFavorite = false;
        try {
            currentUser = UserHelper.getInstance(context).readCurrentUser();
            if (currentUser != null) {
                ObservationFavorite favorite = observation.getFavoritesMap().get(currentUser.getRemoteId());
                isFavorite = favorite != null && favorite.isFavorite();
            }
        } catch (UserException e) {
            Log.e(LOG_NAME, "Could not get user", e);
        }

        return isFavorite;
    }

    private void getDirections(Observation observation) {
        if (observationActionListener != null) {
            observationActionListener.onObservationDirections(observation);
        }
    }

    class IconTask extends AsyncTask<Observation, Void, Bitmap> {
        private final WeakReference<ImageView> reference;

        public IconTask(ImageView imageView) {
            this.reference = new WeakReference<>(imageView);
        }

        @Override
        protected Bitmap doInBackground(Observation... observations) {
            return ObservationBitmapFactory.bitmap(context, observations[0]);
        }

        @Override
        protected void onPostExecute(Bitmap bitmap) {
            if (isCancelled()) {
                bitmap = null;
            }

            ImageView imageView = reference.get();
            if (imageView != null && bitmap != null) {
                imageView.setImageBitmap(bitmap);
            }
        }
    }

    class UserTask extends AsyncTask<Observation, Void, User> {
        private final WeakReference<TextView> reference;

        public UserTask(TextView textView) {
            this.reference = new WeakReference<>(textView);
        }

        @Override
        protected User doInBackground(Observation... observations) {
            User user = null;
            try {
                user = UserHelper.getInstance(context).read(observations[0].getUserId());
            } catch (UserException e) {
                Log.e(LOG_NAME, "Could not get user", e);
            }

            return user;
        }

        @Override
        protected void onPostExecute(User user) {
            if (isCancelled()) {
                user = null;
            }

            TextView textView = reference.get();
            if (textView != null) {
                if (user != null) {
                    textView.setText(user.getDisplayName());
                } else {
                    textView.setText("Unkown User");
                }
            }
        }
    }

    private static class PropertyTask extends AsyncTask<Observation, Void, ObservationProperty> {
        enum Type {
            PRIMARY,
            SECONDARY
        }

        private Type type;
        private final WeakReference<TextView> reference;

        public PropertyTask(Type type, TextView textView) {
            this.type = type;
            this.reference = new WeakReference<>(textView);
        }

        @Override
        protected ObservationProperty doInBackground(Observation... observations) {
            return type == Type.PRIMARY ? observations[0].getPrimaryFeedField() : observations[0].getSecondaryFeedField();
        }

        @Override
        protected void onPostExecute(ObservationProperty property) {
            if (isCancelled()) {
                property = null;
            }

            TextView textView = reference.get();
            if (textView != null) {
                if (property == null || property.isEmpty()) {
                    textView.setVisibility(View.GONE);
                } else {
                    textView.setText(property.getValue().toString());
                    textView.setVisibility(View.VISIBLE);
                }

                textView.requestLayout();
            }
        }
    }
}

File: mage/src/main/java/mil/nga/giat/mage/newsfeed/ObservationListAdapter.kt
package mil.nga.giat.mage.newsfeed

import android.content.Context
import android.database.Cursor
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.GradientDrawable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.chip.Chip
import com.j256.ormlite.android.AndroidDatabaseResults
import com.j256.ormlite.stmt.PreparedQuery
import mil.nga.giat.mage.R
import mil.nga.giat.mage.coordinate.CoordinateFormatter
import mil.nga.giat.mage.database.model.observation.Attachment
import mil.nga.giat.mage.database.model.observation.Observation
import mil.nga.giat.mage.database.model.observation.ObservationFavorite
import mil.nga.giat.mage.data.datasource.observation.ObservationLocalDataSource
import mil.nga.giat.mage.database.model.observation.ObservationImportant
import mil.nga.giat.mage.database.model.observation.ObservationProperty
import mil.nga.giat.mage.map.annotation.MapAnnotation
import mil.nga.giat.mage.observation.attachment.AttachmentCarouselAdapter
import mil.nga.giat.mage.data.datasource.event.EventLocalDataSource
import mil.nga.giat.mage.database.model.user.User
import mil.nga.giat.mage.data.datasource.user.UserLocalDataSource
import mil.nga.giat.mage.sdk.exceptions.ObservationException
import mil.nga.giat.mage.sdk.exceptions.UserException
import mil.nga.giat.mage.utils.DateFormatFactory
import mil.nga.sf.Point
import java.sql.SQLException
import java.util.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class ObservationListAdapter(
   private val context: Context,
   private val userLocalDataSource: UserLocalDataSource,
   private val eventLocalDataSource: EventLocalDataSource,
   private val observationLocalDataSource: ObservationLocalDataSource,
   observationFeedState: ObservationFeedViewModel.ObservationFeedState,
   private val onAttachmentClick: (Attachment) -> Unit,
   private val observationActionListener: ObservationActionListener?,
   private val scope: CoroutineScope
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

   interface ObservationActionListener {
      fun onObservationClick(observation: Observation)
      fun onObservationDirections(observation: Observation)
      fun onObservationLocation(observation: Observation)
   }

   private var cursor: Cursor = observationFeedState.cursor
   private val query: PreparedQuery<Observation> = observationFeedState.query
   private val filterText: String = observationFeedState.filterText
   private var currentUser: User? = null
   private val event = eventLocalDataSource.currentEvent

   private inner class ObservationViewHolder(view: View) : RecyclerView.ViewHolder(view) {
      private val card: View = view.findViewById(R.id.card)
      val markerView: ImageView = view.findViewById(R.id.observation_marker)
      val primaryView: TextView = view.findViewById(R.id.primary)
      val secondaryView: TextView = view.findViewById(R.id.secondary)
      val userView: TextView = view.findViewById(R.id.user)
      val importantView: View = view.findViewById(R.id.important)
      val importantOverline: TextView = view.findViewById(R.id.important_overline)
      val importantDescription: TextView = view.findViewById(R.id.important_description)
      val syncBadge: View = view.findViewById(R.id.sync_status)
      val errorBadge: View = view.findViewById(R.id.error_status)
      val attachmentCarousel: RecyclerView = view.findViewById(R.id.attachment_carousel)
      val attachmentDots: LinearLayout = view.findViewById(R.id.attachment_dots)
      val attachmentPageCount: Chip = view.findViewById(R.id.attachment_page_count)
      val attachmentFailedBadge: Chip = view.findViewById(R.id.attachment_failed_badge)
      var dotViews: List<View> = emptyList()
      val locationView: TextView = view.findViewById(R.id.location)
      val locationContainer: View = view.findViewById(R.id.location_container)
      val favoriteButton: ImageView = view.findViewById(R.id.favorite_button)
      val favoriteCount: TextView = view.findViewById(R.id.favorite_count)
      val directionsButton: View = view.findViewById(R.id.directions_button)
      val timeView: TextView = view.findViewById(R.id.time)

      val attachmentCarouselAdapter = AttachmentCarouselAdapter(context, onAttachmentClick)
      var totalAttachmentCount = 0

      var timestamp: Date? = null
      var centroid: Point? = null
      var bindJob: Job? = null

      init {
         attachmentCarousel.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
         attachmentCarousel.adapter = attachmentCarouselAdapter
         PagerSnapHelper().attachToRecyclerView(attachmentCarousel)
         attachmentCarousel.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
               val layoutManager = recyclerView.layoutManager as? LinearLayoutManager ?: return
               val position = layoutManager.findFirstVisibleItemPosition()
               if (position != RecyclerView.NO_POSITION && totalAttachmentCount > 1) {
                  attachmentPageCount.text = "${position + 1} of $totalAttachmentCount"
                  updateActiveDot(position)
               }
            }
         })
      }

      fun setupDots(count: Int) {
         attachmentDots.removeAllViews()
         if (count <= 1) {
            attachmentDots.visibility = View.GONE
            dotViews = emptyList()
            return
         }
         attachmentDots.visibility = View.VISIBLE
         val dotSize = (6 * context.resources.displayMetrics.density).toInt()
         val dotMargin = (3 * context.resources.displayMetrics.density).toInt()
         dotViews = (0 until count).map { index ->
            val dot = View(context)
            val params = LinearLayout.LayoutParams(dotSize, dotSize)
            params.setMargins(dotMargin, 0, dotMargin, 0)
            dot.layoutParams = params
            dot.background = dotDrawable(index == 0)
            attachmentDots.addView(dot)
            dot
         }
      }

      fun updateActiveDot(position: Int) {
         dotViews.forEachIndexed { index, dot ->
            dot.background = dotDrawable(index == position)
         }
      }

      private fun dotDrawable(active: Boolean): GradientDrawable {
         return GradientDrawable().apply {
            shape = GradientDrawable.OVAL
            setColor(if (active) Color.WHITE else Color.argb(102, 255, 255, 255))
         }
      }

      fun bind(observation: Observation) {
         timestamp = observation.timestamp
         centroid = observation.geometry.centroid

         card.setOnClickListener { observationActionListener?.onObservationClick(observation) }
      }
   }

   private inner class FooterViewHolder(view: View) :
      RecyclerView.ViewHolder(view) {
      val footerText: TextView = view.findViewById(R.id.footer_text)

   }

   override fun getItemViewType(position: Int): Int {
      return if (position == cursor.count) {
         TYPE_FOOTER
      } else {
         TYPE_OBSERVATION
      }
   }

   override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
      return if (viewType == TYPE_OBSERVATION) {
         val itemView = LayoutInflater.from(parent.context).inflate(R.layout.observation_list_item, parent, false)
         ObservationViewHolder(itemView)
      } else {
         val itemView = LayoutInflater.from(parent.context).inflate(R.layout.feed_footer, parent, false)
         FooterViewHolder(itemView)
      }
   }

   override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
      when (holder) {
         is ObservationViewHolder -> bindObservation(holder, position)
         else -> bindFooter(holder)
      }
   }

   override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int, payloads: List<Any>) {
      if (payloads.isEmpty()) {
         super.onBindViewHolder(holder, position, payloads)
      } else {
         if (holder is ObservationViewHolder) {
            for (payload in payloads) {
               if (payload == PAYLOAD_TIMEZONE_CHANGE) {
                  updateTimeZoneDisplay(holder)
               }
               if (payload == PAYLOAD_COORDINATE_CHANGE) {
                  updateCoordinateDisplay(holder)
               }
            }
         }
      }
   }

   private fun updateTimeZoneDisplay(vh: ObservationViewHolder) {
      vh.timestamp?.let { timestamp ->
         val dateFormat = DateFormatFactory.format("yyyy-MM-dd HH:mm zz", Locale.getDefault(), context)
         vh.timeView.text = dateFormat.format(timestamp)
      }
   }

   private fun updateCoordinateDisplay(vh: ObservationViewHolder) {
      vh.centroid?.let { point ->
         val coordinates = CoordinateFormatter(context).format(LatLng(point.y, point.x))
         vh.locationView.text = coordinates
      }
   }

   override fun getItemCount(): Int {
      return cursor.count + 1
   }

   override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
      if (holder is ObservationViewHolder) {
          holder.bindJob?.cancel()
      }
   }


   private fun bindObservation(holder: RecyclerView.ViewHolder, position: Int) {
      cursor.moveToPosition(position)

      val vh = holder as ObservationViewHolder

      try {
         val observation = query.mapRow(AndroidDatabaseResults(cursor, null, false))
         vh.bind(observation)

         val markerPlaceholder = DrawableCompat.wrap(ContextCompat.getDrawable(context, R.drawable.ic_place_white_48dp)!!)
         DrawableCompat.setTint(markerPlaceholder, ContextCompat.getColor(context, R.color.icon))
         DrawableCompat.setTintMode(markerPlaceholder, PorterDuff.Mode.SRC_IN)
         vh.markerView.setImageDrawable(markerPlaceholder)

         val observationForm = observation.forms.firstOrNull()
         val formDefinition = observationForm?.formId?.let { formId ->
            eventLocalDataSource.getForm(formId)
         }

         val icon = MapAnnotation.getAnnotationWithStyleFromObservation(
            event = event,
            observation = observation,
            formDefinition = formDefinition,
            observationForm = observationForm,
            geometryType = observation.geometry.geometryType,
            context = context
         )

         Glide.with(context)
            .asBitmap()
            .load(icon)
            .error(R.drawable.default_marker)
            .into(vh.markerView)

         vh.primaryView.text = ""
         vh.secondaryView.text = ""
         vh.userView.text = "" 

         // Cancel any in-flight load from a previous binding of this ViewHolder
         vh.bindJob?.cancel()
         vh.bindJob = scope.launch {
               // Fetch user name from local DB on the IO thread
               val user = withContext(Dispatchers.IO) {
                  try { userLocalDataSource.read(observation.userId) } catch (e: Exception) { null }
               }   
               
               // Load the form definition once — primary and secondary fields both need it
               val observationForm = observation.forms.firstOrNull()
               val form = observationForm?.let {
                  withContext(Dispatchers.IO) { eventLocalDataSource.getForm(it.formId) }
               }   
               
               // Property look-ups are in-memory once the form is loaded
               val primaryProperty = observationForm?.properties?.find { it.key == form?.primaryFeedField }
               val secondaryProperty = observationForm?.properties?.find { it.key == form?.secondaryFeedField }
               
               val importantUserName = withContext(Dispatchers.IO) {
                  try {
                     observation.important?.userId?.let { userLocalDataSource.read(it)?.displayName }
                  } catch (e: Exception) { null }
               }

               // Everything below runs back on the main thread (scope is Main)
               vh.userView.text = user?.displayName ?: "Unknown User"

               if (primaryProperty == null || primaryProperty.isEmpty) {
                  vh.primaryView.visibility = View.GONE
               } else {
                  vh.primaryView.text = primaryProperty.value.toString()
                  vh.primaryView.visibility = View.VISIBLE
               }
               vh.primaryView.requestLayout()

               if (secondaryProperty == null || secondaryProperty.isEmpty) {
                  vh.secondaryView.visibility = View.GONE
               } else {
                  vh.secondaryView.text = secondaryProperty.value.toString()
                  vh.secondaryView.visibility = View.VISIBLE
               }
               vh.secondaryView.requestLayout()

               setImportantView(observation.important, importantUserName, vh)
         }

         updateTimeZoneDisplay(vh)

         val error = observation.error
         if (error != null) {
            vh.errorBadge.visibility = if (error.statusCode != null) View.VISIBLE else View.GONE
         } else {
            vh.syncBadge.visibility = if (observation.isDirty) View.VISIBLE else View.GONE
            vh.errorBadge.visibility = View.GONE
         }

         vh.attachmentCarouselAdapter.submitAttachments(observation.attachments)
         vh.totalAttachmentCount = observation.attachments.size
         vh.attachmentCarousel.scrollToPosition(0)
         vh.setupDots(observation.attachments.size)
         if (observation.attachments.size > 1) {
            vh.attachmentPageCount.text = "1 of ${observation.attachments.size}"
            vh.attachmentPageCount.visibility = View.VISIBLE
         } else {
            vh.attachmentPageCount.visibility = View.GONE
         }

         val failedAttachmentCount = observation.attachments.count {
            it.processingStatus == "rejected" || it.processingStatus == "error"
         }
         if (failedAttachmentCount > 0) {
            vh.attachmentFailedBadge.text = if (failedAttachmentCount > 1) "Attachments Failed - $failedAttachmentCount" else "Attachment Failed"
            vh.attachmentFailedBadge.visibility = View.VISIBLE
         } else {
            vh.attachmentFailedBadge.visibility = View.GONE
         }

         updateCoordinateDisplay(vh)
         vh.locationContainer.setOnClickListener { onLocationClick(observation) }

         vh.favoriteButton.setOnClickListener { toggleFavorite(observation, vh) }
         setFavoriteImage(observation.favorites, vh, isFavorite(observation))

         vh.directionsButton.setOnClickListener { getDirections(observation) }
      } catch (e: SQLException) {
         Log.e(LOG_NAME, "Error reading observation from database", e)
      }
   }

   private fun bindFooter(holder: RecyclerView.ViewHolder) {
      val vh = holder as FooterViewHolder
      var footerText = "End of results"
      if (filterText.isNotEmpty()) {
         footerText = "End of results for $filterText"
      }
      vh.footerText.text = footerText
   }

   private fun setImportantView(important: ObservationImportant?, flaggedByName: String?, vh: ObservationViewHolder) {
      val isImportant = important != null && important.isImportant
      vh.importantView.visibility = if (isImportant) View.VISIBLE else View.GONE
      if (isImportant) {
         vh.importantOverline.text = String.format("FLAGGED BY %s", flaggedByName?.uppercase(Locale.getDefault()) ?: "UNKNOWN")
         vh.importantDescription.text = important!!.description
      }
   }

   private fun setFavoriteImage(favorites: Collection<ObservationFavorite>, vh: ObservationViewHolder, isFavorite: Boolean) {
      if (isFavorite) {
         vh.favoriteButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_favorite_white_24dp))
         vh.favoriteButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_favorite_active))
      } else {
         vh.favoriteButton.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_favorite_border_white_24dp))
         vh.favoriteButton.setColorFilter(ContextCompat.getColor(context, R.color.observation_favorite_inactive))
      }

      vh.favoriteCount.visibility = if (favorites.isNotEmpty()) View.VISIBLE else View.GONE
      vh.favoriteCount.text = String.format(Locale.getDefault(), "%d", favorites.size)
   }

   private fun toggleFavorite(observation: Observation, vh: ObservationViewHolder) {
      val isFavorite = isFavorite(observation)
      userLocalDataSource.readCurrentUser()?.let { user ->
         try {
            if (isFavorite) {
               observationLocalDataSource.unfavoriteObservation(observation, user)
            } else {
               observationLocalDataSource.favoriteObservation(observation, user)
            }
            setFavoriteImage(observation.favorites, vh, isFavorite)
         } catch (e: ObservationException) {
            Log.e(LOG_NAME, "Could not unfavorite observation", e)
         }
      }
   }

   private fun isFavorite(observation: Observation): Boolean {
      var isFavorite = false
      try {
         userLocalDataSource.readCurrentUser()?.let { user ->
            currentUser = user
            val favorite = observation.favoritesMap[user.remoteId]
            isFavorite = favorite != null && favorite.isFavorite
         }
      } catch (e: UserException) {
         Log.e(LOG_NAME, "Could not get user", e)
      }
      return isFavorite
   }

   private fun getDirections(observation: Observation) {
      observationActionListener?.onObservationDirections(observation)
   }

   private fun onLocationClick(observation: Observation) {
      observationActionListener?.onObservationLocation(observation)
   }

   companion object {
      private val LOG_NAME = ObservationListAdapter::class.java.name
      private const val TYPE_OBSERVATION = 1
      private const val TYPE_FOOTER = 2
      const val PAYLOAD_TIMEZONE_CHANGE = "PAYLOAD_TIMEZONE_CHANGE"
      const val PAYLOAD_COORDINATE_CHANGE = "PAYLOAD_COORDINATE_CHANGE"
   }
}